File: samplesJS/Animations/src/AnimationsSample.cpp
// Required for WinMain
#include <Windows.h>

// Oak Nut include
#include <onut/Anim.h>
#include <onut/onut.h>
#include <onut/Renderer.h>
#include <onut/Settings.h>
#include <onut/Sound.h>
#include <onut/SpriteBatch.h>
#include <onut/Texture.h>

void init();
void update();
void render();

// Main
int CALLBACK WinMain(HINSTANCE appInstance, HINSTANCE prevInstance, LPSTR cmdLine, int cmdCount)
{
    oSettings->setGameName("Animations Sample");
    ORun(init, update, render);
}

OAnimFloat angleAnim1;
OAnimFloat angleAnim2;
OAnimVector2 bounceAnim;
OAnimRect frameAnim;
OAnimVector2 callbackAnim;

void playBounceSound()
{
    OPlaySound("bounce.wav", .5f);
}

void init()
{
    // Loop
    angleAnim1.play(
        0.f, // From
        360.f, // To
        2.f, // Duration in Seconds
        OTweenLinear, // Tween
        OLoop // Loop
        );

    // Pingpong loop
    angleAnim2.play(
        0.f, // From
        720.f, // To
        2.f, // Duration in Seconds
        OTweenEaseBoth, // Tween
        OPingPongLoop // Loop
        );
    
    // Bounce
    bounceAnim.play(
        Vector2(0.f, 0.f), // From
        Vector2(0.f, 200.f), // To
        2.f, // Duration in Seconds
        OTweenBounceOut, // Tween
        OLoop // Loop
        );
    
    // Keyframes
    frameAnim.playKeyFrames(
        Rect(32.f, 192.f, 128.f, 128.f), // From
        {   
            // Keyframes
            {Rect(32.f, 192.f, 128.f, 400.f), 2.f, OTweenEaseBoth},
            {Rect(32.f, 192.f, 192.f, 400.f), 1.5f, OTweenBounceOut},
            {Rect(32.f, 192.f, 128.f, 128.f), 1.5f, OTweenBounceOut},
        }, 
        OLoop // Loop
        );
    
    // Queuing + Callbacks
    callbackAnim = Vector2::Zero;
    callbackAnim.queue(Vector2(256, 0), 1.5f, OTweenEaseIn, playBounceSound);
    callbackAnim.queue(Vector2::Zero, 1.5f, OTweenEaseIn, playBounceSound);
    callbackAnim.play(OLoop);
}

void update()
{
}

void render()
{
    auto pNutTexture = OGetTexture("onutLogo.png");
    auto pFrameTexture = OGetTexture("frameSmall.png");

    // Clear
    oRenderer->clear(OColorHex(1d232d));

    // Draw sprites
    oSpriteBatch->begin();
    oSpriteBatch->drawSprite(pNutTexture, Vector2(128, 128), Color::White, angleAnim1);
    oSpriteBatch->drawSprite(pNutTexture, Vector2(256, 128), Color::White, angleAnim2);
    oSpriteBatch->drawSprite(pNutTexture, Vector2(384, 128) + bounceAnim, Color::White);
    oSpriteBatch->drawRectScaled9RepeatCenters(pFrameTexture, frameAnim, Vector4(48, 48, 48, 48));
    oSpriteBatch->drawSprite(pNutTexture, Vector2(512, 256) + callbackAnim);
    oSpriteBatch->end();
}

File: samplesJS/Animations/assets/scripts/AnimationsSample.js
var nut = getTexture("onutLogo.png");
var frame = getTexture("frameSmall.png");

var angleAnim1 = new NumberAnim();
var angleAnim2 = new NumberAnim();
var bounceAnim = new Vector2Anim();
var frameAnim = new RectAnim();
var callbackAnim = new Vector2Anim();
var batchTransform = new MatrixAnim();

function playBounceSound() {
    playSound("bounce.wav", .5);
}

// Loop
angleAnim1.playSingle(0, 360, 2, Tween.LINEAR, Loop.LOOP);

// Pingpong loop
angleAnim2.playSingle(0, 720, 2, Tween.EASE_BOTH, Loop.PING_PONG_LOOP);

// Bounce
bounceAnim.playSingle(new Vector2(0, 0), new Vector2(0, 200), 2, Tween.BOUNCE_OUT, Loop.LOOP);

// Keyframes
frameAnim.playKeyFrames(new Rect(32, 192, 128, 128), [
    { value: new Rect(32, 192, 128, 400), duration: 2, tween: Tween.EASE_BOTH },
    { value: new Rect(32, 192, 192, 400), duration: 1.5, tween: Tween.BOUNCE_OUT },
    { value: new Rect(32, 192, 128, 128), duration: 1.5, tween: Tween.BOUNCE_OUT },
], Loop.LOOP);

// Queuing + Callbacks
callbackAnim.set(Vector2.ZERO);
callbackAnim.queue(new Vector2(256, 0), 1.5, Tween.EASE_IN, playBounceSound);
callbackAnim.queue(Vector2.ZERO, 1.5, Tween.EASE_IN, playBounceSound);
callbackAnim.play(Loop.LOOP);

function update(dt) {
    // Create some crazy screen animation using a Matrix transform.
    if (Input.isJustDown(Key.SPACE_BAR)) {
        var screen = Renderer.getResolution();
        batchTransform.playKeyFrames(Matrix.IDENTITY, [
            {
                value: Matrix.createTranslation(new Vector3(-screen.x, -screen.y, 0)).mul(
                    Matrix.createRotationZ(45).mul(
                        Matrix.createTranslation(new Vector3(screen.x, screen.y, 0)))),
                duration: 1, tween: Tween.LINEAR
            },
            {
                value: Matrix.createTranslation(new Vector3(-screen.x, -screen.y, 0)).mul(
                    Matrix.createRotationZ(-45).mul(
                        Matrix.createTranslation(new Vector3(screen.x, screen.y, 0)))),
                duration: 2, tween: Tween.LINEAR
            },
            {
                value: Matrix.IDENTITY, duration: 1, tween: Tween.LINEAR
            }
        ]);
    }
}

function render() {
    // Clear
    Renderer.clear(Color.fromHexRGB(0x1d232d));

    // Draw sprites
    SpriteBatch.begin(batchTransform.get());
    SpriteBatch.drawSprite(nut, new Vector2(128, 128), Color.WHITE, angleAnim1.get());
    SpriteBatch.drawSprite(nut, new Vector2(256, 128), Color.WHITE, angleAnim2.get());
    SpriteBatch.drawSprite(nut, new Vector2(384, 128).add(bounceAnim.get()), Color.WHITE);
    SpriteBatch.drawRectScaled9RepeatCenters(frame, frameAnim.get(), new Vector4(48, 48, 48, 48));
    SpriteBatch.drawSprite(nut, new Vector2(512, 256).add(callbackAnim.get()));
    SpriteBatch.end();
}
